docs/rfcs/0011-customizable-store-dispatcher.md

```diff
@@ -0,0 +1,71 @@
+
+When components dispatch actions to the store, these actions are performed on an independant single thread dispatcher in the `Store` to avoid overloading the main thread with heavy work that might be performed during the `reduce` or in a `Middleware`. In practice, these actions have been short and fast so they do not cause overhead. Work done in a `Middleware` that can be slow, like I/O,  is put onto separate Dispatchers. The performance optimization to switch to a `Store` thread, requires that components which are always run on the main thread, to ensure synchronisation is now kept between the main thread and the store thread for observers of the `State`.
+
+There are some advantages to this change:
+
+* Simplicity for main-thread `Store`s.
+* Unit testing can now occur on the test framework's thread.
+* Fewer resources needed for context shifting between threads.
+
+## Guide-level explanation
+
+There are two possible ways to allow for this behaviour during the store creation:
+
+1. Pass in a `CoroutineDispatcher`.
+2. Configurable set of pre-defined options to choose.
+
+Passing in a `CoroutineDispatcher` has the flexibility to allow any preferred dispatcher to be used (e.g. `Dispatchers.IO` or Android Component's `Dispatchers.Cached`),  however that could lead to complications on choosing the wrong dispatcher. Ensuring that testing is also done with the same dispatcher configuration can also be difficult with this pattern.
+
+With a predefined set of options, we have a smaller surface area of changes that we want to maintain, possibly starting off with either the main dispatcher or the current single threaded dispatcher. We can continue to use our `StoreThreadFactory` and disable it when we switch to the `Main` dispatcher.
+
+Below are examples of what either surface would look like to a consumer of an "AppStore":
+
+```kotlin
+// Pass in a `CoroutineDispatcher`.
+class AppStore(
+    initialState: AppState = AppState(),
+) : Store<AppState, AppAction>(
+    initialState = initialState,
+    reducer = AppStoreReducer::reduce,
+    dispatcher = Dispatchers.Main // choose a preferable dispatcher.
+)
+
+// Configurable set of pre-defined options to choose.
+class AppStore(
+    initialState: AppState = AppState(),
+) : Store<AppState, AppAction>(
+    initialState = initialState,
+    reducer = AppStoreReducer::reduce,
+    runOnMain = true, // a boolean or enum value to signify the dispatcher used.
+)
+```
+
+Another advantage of this change is that `Store`s would be able to run on the test framework's default dispatcher. This reduces the number of intermittent test failures we have seen where the `Store` is still processing actions async to an assertion. Fewer usages of `store.waitUntilIdle()` are required after every `dispatch` is called in a test, as a result.
```

Comment:
I'm not sure I understand what you are asking here in regards to this RFC.

> If we are only concerned about the result of the reducer, then `store.dispatch().joinBlocking()` should suffice and can be used across threads.

From my understanding, `joinBlocking` will block the running of the test for that particular action alone and then continue with that test. With `waitUntilIdle`, we would wait until all work with that Store scope would be completed, so actions that are dispatched as side-effects of the original one. 

You can use `waitUntilIdle` in all cases today, but you can't use `joinBlocking` for all cases that `waitUntilIdle` is used.

> this change would not be sufficient to alleviate this concern without allowing the middleware to inherit the dispatcher of the Store.

Isn't this something that already exists if true? We are not trying to make sure all Middlewares inherit the same scope. You can always pass in a separate scope or dispatcher for work that you don't want to be done in the Store scope.

Reply:
> From my understanding, joinBlocking will block the running of the test for that particular action alone and then continue with that test. With waitUntilIdle, we would wait until all work with that Store scope would be completed, so actions that are dispatched as side-effects of the original one.

Yeah, this is exactly right. So `joinBlocking` should suffice for any case where we are testing just the reducer, whereas `waitUntilIdle` is required if, for example, middlewares dispatch additional actions. However, `waitUntilIdle` doesn't account for coroutines launched by the middleware. 

> Fewer usages of `store.waitUntilIdle()` are required after every `dispatch` is called in a test, as a result.

Wouldn't we still need as many `waitUntilIdle`s (or `testScope.advanceUntilIdle`s) because the processing is still happening async, just on the test thread instead of a Store thread?

Reply:
> Wouldn't we still need as many `waitUntilIdle`s (or `testScope.advanceUntilIdle`s) because the processing is still happening async, just on the test thread instead of a Store thread?

In our tests, we use the `MainCoroutineRule` so that all main tasks are done on the same test thread, so that would make all main work blocking in the sequence that they are executed:

```kotlin
@get:Rule
val coroutinesTestRule = MainCoroutineRule()
```

I will update the RFC to mention this because it isn't explicit.

Reply:
> In our tests, we use the MainCoroutineRule so that all main tasks are done on the same test thread, so that would make all main work blocking in the sequence that they are executed:

This is tangential, but just for clarity the problem I would be interested in solving is when not all tasks are done on the main thread. For example, I don't think this test will pass even after this proposal:

```kotlin
@Test
    fun `middleware does not finish executing`() = runTestOnMain {
        var captured = false
        val middleware = object : Middleware<TestState, TestAction> {
            override fun invoke(
                context: MiddlewareContext<TestState, TestAction>,
                next: (TestAction) -> Unit,
                action: TestAction
            ) {
                when (action) {
                    TestAction.IncrementAction -> this@runTestOnMain.launch(Dispatchers.IO) {
                        delay(5000)
                        captured = true
                    }
                    else -> Unit
                }
            }
        }
        val store = Store(
            TestState(0),
            ::reducer,
            listOf(middleware),
        )
        store.dispatch(TestAction.IncrementAction).joinBlocking()

        assertTrue(captured)
    }
```

Reply:
That's correct, we are not solving for that use case. If the developer is providing their own dispatcher for the side-effects, it's up to them to pass that in through the constructor ([example](https://searchfox.org/mozilla-mobile/rev/98a241402bf6ecfa7fdd302006604393e02b9d50/firefox-android/android-components/components/feature/downloads/src/test/java/mozilla/components/feature/downloads/DownloadMiddlewareTest.kt#56)) or other means for the test.